In [11]:
import pandas as pd 

data_path = 'seoul_std_link_service_link_mapping_20250306.xlsx'
df = pd.read_excel(data_path)
print(df.head())


        서비스링크  표준링크순서     표준링크아이디
0  1000000100     1.0  1000022501
1  1000000100     2.0  1000022502
2  1000000200     1.0  1000022401
3  1000000200     2.0  1000022402
4  1000000300     3.0  1000000303


In [12]:
df = df.rename(columns={'서비스링크': 'service_link_id', '표준링크순서': 'std_link_order', '표준링크아이디': 'std_link_id'})
print(df.head())

   service_link_id  std_link_order  std_link_id
0       1000000100             1.0   1000022501
1       1000000100             2.0   1000022502
2       1000000200             1.0   1000022401
3       1000000200             2.0   1000022402
4       1000000300             3.0   1000000303


In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11535 entries, 0 to 11534
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   service_link_id  11535 non-null  int64  
 1   std_link_order   11512 non-null  float64
 2   std_link_id      11535 non-null  int64  
dtypes: float64(1), int64(2)
memory usage: 270.5 KB


In [21]:
df.groupby('service_link_id').agg({
    'link_id': ['count', 'nunique'],
    'service_link_id': ['count', 'nunique']
})

link_id         service_link_id        
                  count nunique           count nunique
service_link_id                                        
1000000301            1       1               1       1
1000000302            1       1               1       1
1000000303            1       1               1       1
1000000401            1       1               1       1
1000000402            1       1               1       1
...                 ...     ...             ...     ...
2270545700            1       1               1       1
2270545800            1       1               1       1
2270545900            1       1               1       1
2320127400            1       1               1       1
2320127500            1       1               1       1

[11535 rows x 4 columns]

In [ ]:
import pandas as pd
import geopandas as gpd
import folium
import pandas as pd
from shapely.geometry import MultiLineString, LineString

# link.shp 파일 읽기 , 전국
link_df=gpd.read_file("./20250512_nodelinkdata/MOCT_LINK.shp",encoding='euc-kr')
link_df['geometry'] = link_df['geometry'].to_crs('EPSG:4326') # 위도경도 좌표계 변환

In [ ]:
# Node.shp 파일읽기. 전국 노드 정보 
nation_wide_node_df = gpd.read_file("./20250512_nodelinkdata/MOCT_NODE.shp",encoding='euc-kr')
nation_wide_node_df['geometry'] = nation_wide_node_df['geometry'].to_crs('EPSG:4326') # 위도경도 좌표계 변환

In [44]:
nation_wide_node_df.columns

Index(['NODE_ID', 'NODE_TYPE', 'NODE_NAME', 'TURN_P', 'UPDATEDATE', 'REMARK',
       'HIST_TYPE', 'HISTREMARK', 'geometry'],
      dtype='object')

In [45]:
temp=nation_wide_node_df['NODE_ID'].apply(lambda x:int(x[:3]))
# 서울 코드 노드만 추출 
seoul_node_df=nation_wide_node_df[(temp>=100)&(temp<=124)] ### 서울 코드 100~124
seoul_node_df

,NODE_ID,NODE_TYPE,NODE_NAME,TURN_P,UPDATEDATE,REMARK,HIST_TYPE,HISTREMARK,geometry
2,1020134000,101,한강로2가 412,0,20250512,None,NODE0001,None,POINT (126.96939 37.52969)
4,1150266000,101,개화동 550-6,0,20250512,None,NODE0001,None,POINT (126.80074 37.5777)
81,1249001500,104,교량서측,0,20241227,None,NODE0001,None,POINT (127.15017 37.56543)
102,1190025605,101,신호빌딩,0,20250512,None,NODE0001,None,POINT (126.9222 37.4977)
171,1160120900,101,구로동 462-1,0,20250512,None,NODE0001,None,POINT (126.88252 37.49598)
...,...,...,...,...,...,...,...,...,...
1173458,1050148100,101,이문동 425,0,20250512,None,NODE0001,None,POINT (127.05629 37.60002)
1173459,1120021004,101,홍은교,0,20250512,None,NODE0001,None,POINT (126.9459 37.59452)
1173481,1249000700,107,최소노드배치점,0,20241227,None,NODE0001,None,POINT (127.15693 37.55495)
1173594,1000015901,101,중앙지하차도(서측),0,20241129,None,NODE0005,None,POINT (126.97563 37.57411)


In [2]:
link_df.columns

Index(['LINK_ID', 'F_NODE', 'T_NODE', 'LANES', 'ROAD_RANK', 'ROAD_TYPE',
       'ROAD_NO', 'ROAD_NAME', 'ROAD_USE', 'MULTI_LINK', 'CONNECT', 'MAX_SPD',
       'REST_VEH', 'REST_W', 'REST_H', 'C-ITS', 'LENGTH', 'UPDATEDATE',
       'REMARK', 'HIST_TYPE', 'HISTREMARK', 'geometry'],
      dtype='object')

In [25]:
link_df.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
RangeIndex: 1548963 entries, 0 to 1548962
Data columns (total 22 columns):
 #   Column      Non-Null Count    Dtype   
---  ------      --------------    -----   
 0   LINK_ID     1548963 non-null  object  
 1   F_NODE      1548963 non-null  object  
 2   T_NODE      1548963 non-null  object  
 3   LANES       1548963 non-null  int32   
 4   ROAD_RANK   1548963 non-null  object  
 5   ROAD_TYPE   1548963 non-null  object  
 6   ROAD_NO     1548963 non-null  object  
 7   ROAD_NAME   1548963 non-null  object  
 8   ROAD_USE    1548963 non-null  object  
 9   MULTI_LINK  1548963 non-null  object  
 10  CONNECT     1548963 non-null  object  
 11  MAX_SPD     1548963 non-null  int32   
 12  REST_VEH    1548963 non-null  object  
 13  REST_W      1548963 non-null  int32   
 14  REST_H      1548963 non-null  int32   
 15  C-ITS       1548963 non-null  object  
 16  LENGTH      1548963 non-null  float64 
 17  UPDATEDATE  1548963 non-null  object  

In [ ]:
temp=link_df['LINK_ID'].apply(lambda x:int(x[:3]))
# 서울 코드 링크만 추출 
slink_df=link_df[(temp>=100)&(temp<=124)] ### 서울 코드 100~124

# 시각화 함수
def linechart(df):
    m = folium.Map(location=[37.5665, 126.9780], zoom_start=15)
    for _, row in df.iterrows():
        if isinstance(row['geometry'], MultiLineString) and len(row['geometry'].geoms) > 1:
            for line_string in row['geometry'].geoms:
                coordinates = [(lat, lon) for lon, lat in line_string.coords]
                folium.PolyLine(locations=coordinates, color='blue').add_to(m)
        else:
            coordinates = [(lat, lon) for lon, lat in zip(*row['geometry'].xy)]
            folium.PolyLine(locations=coordinates, color='blue').add_to(m)
    return m



In [ ]:
# linechart(slink_df)

In [9]:
print(slink_df.columns)

Index(['LINK_ID', 'F_NODE', 'T_NODE', 'LANES', 'ROAD_RANK', 'ROAD_TYPE',
       'ROAD_NO', 'ROAD_NAME', 'ROAD_USE', 'MULTI_LINK', 'CONNECT', 'MAX_SPD',
       'REST_VEH', 'REST_W', 'REST_H', 'C-ITS', 'LENGTH', 'UPDATEDATE',
       'REMARK', 'HIST_TYPE', 'HISTREMARK', 'geometry'],
      dtype='object')


In [24]:
slink_df.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Index: 63288 entries, 51 to 1548959
Data columns (total 22 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   LINK_ID     63288 non-null  object  
 1   F_NODE      63288 non-null  object  
 2   T_NODE      63288 non-null  object  
 3   LANES       63288 non-null  int32   
 4   ROAD_RANK   63288 non-null  object  
 5   ROAD_TYPE   63288 non-null  object  
 6   ROAD_NO     63288 non-null  object  
 7   ROAD_NAME   63288 non-null  object  
 8   ROAD_USE    63288 non-null  object  
 9   MULTI_LINK  63288 non-null  object  
 10  CONNECT     63288 non-null  object  
 11  MAX_SPD     63288 non-null  int32   
 12  REST_VEH    63288 non-null  object  
 13  REST_W      63288 non-null  int32   
 14  REST_H      63288 non-null  int32   
 15  C-ITS       63288 non-null  object  
 16  LENGTH      63288 non-null  float64 
 17  UPDATEDATE  63288 non-null  object  
 18  REMARK      860 non-null    object  
 19

네가 설명한 구조는 이거잖아:

df: 서울시 서비스링크 테이블
service_link_id
std_link_order
std_link_id
slink_df: 전국 도로 링크 중 서울 코드만 뽑은 테이블
LINK_ID
geometry 등 도로 속성

그러면 매칭 기준은:

df['std_link_id'] 와 slink_df['LINK_ID']

이야.

1. left

왼쪽 데이터프레임

2. right

오른쪽 데이터프레임

3. left_on='컬럼명'

왼쪽 데이터프레임에서 어떤 컬럼을 기준으로 합칠지

4. right_on='컬럼명'

오른쪽 데이터프레임에서 어떤 컬럼을 기준으로 합칠지

5. how='inner'

양쪽에 둘 다 존재하는 키만 남김

In [21]:
# merge 전 컬럼 데이터 타입 맞춰주기 
df['std_link_id'] = df['std_link_id'].astype(str).str.strip()
slink_df['LINK_ID'] = slink_df['LINK_ID'].astype(str).str.strip()
# merge 
slink_df1=pd.merge(left=df,right=slink_df, left_on='std_link_id', right_on = 'LINK_ID',how='inner')

/opt/homebrew/anaconda3/envs/yolo/lib/python3.10/site-packages/geopandas/geodataframe.py:1969: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  super().__setitem__(key, value)


In [23]:
slink_df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11011 entries, 0 to 11010
Data columns (total 25 columns):
 #   Column           Non-Null Count  Dtype   
---  ------           --------------  -----   
 0   service_link_id  11011 non-null  int64   
 1   std_link_order   10994 non-null  float64 
 2   std_link_id      11011 non-null  object  
 3   LINK_ID          11011 non-null  object  
 4   F_NODE           11011 non-null  object  
 5   T_NODE           11011 non-null  object  
 6   LANES            11011 non-null  int32   
 7   ROAD_RANK        11011 non-null  object  
 8   ROAD_TYPE        11011 non-null  object  
 9   ROAD_NO          11011 non-null  object  
 10  ROAD_NAME        11011 non-null  object  
 11  ROAD_USE         11011 non-null  object  
 12  MULTI_LINK       11011 non-null  object  
 13  CONNECT          11011 non-null  object  
 14  MAX_SPD          11011 non-null  int32   
 15  REST_VEH         11011 non-null  object  
 16  REST_W           11011 non-null  int32  

In [35]:
slink_df1['ROAD_NO'].unique() # road no이 - 로 되어 있는 게 있음. 

array(['6', '30', '-', '48', '70', '47', '61', '3', '43', '46', '44', '1',
       '371', '120', '88', '39', '55', '397', '351', '94', '309', '51',
       '23', '342', '60'], dtype=object)

In [37]:
slink_df1['ROAD_USE'].unique()

array(['0', '1'], dtype=object)

* 전국의 1548963개 (1.6 백만)개 도로 중, 63288개 (6만 3천개)의 도로가 서울 코드로 분류되어 있으며 (서울 코드 100~124),
이 정보와 서울 TOPIS에서 주요 도로 정보들에 대한 API 정보가 제공되는 도로 수는 11011개 (1.1 만개) 이다. 
* 서울 TOPIS API에서 도로 개수는 11535 개로 전국 도로 링크에 대한 정보에는 524 (11535 - 11011)의 도로 정보가 없다.